Source/santasyncservice/SNTSyncEventUpload.mm

```diff
@@ -91,18 +80,40 @@ - (BOOL)uploadEvents:(NSArray<SNTStoredEvent *> *)events {
             addObject:santa::StringToNSString(bundle_binary)];
       }
     }
-    SLOGI(@"Uploaded %d events", uploadEvents->size());
+    SLOGI(@"Uploaded %d events", req->events_size());
   }
+  return YES;
+}
 
-  // Remove event IDs. For Bundle Events the ID is 0 so nothing happens.
-  [[self.daemonConn remoteObjectProxy] databaseRemoveEventsWithIDs:[eventIds allObjects]];
+- (BOOL)uploadEvents:(NSArray<SNTStoredEvent *> *)events {
+  google::protobuf::Arena arena;
+  NSMutableSet *eventIds = [NSMutableSet setWithCapacity:events.count];
+  auto req = google::protobuf::Arena::Create<::pbv1::EventUploadRequest>(&arena);
+  req->set_machine_id(NSStringToUTF8String(self.syncState.machineID));
+  google::protobuf::RepeatedPtrField<::pbv1::Event> *uploadEvents = req->mutable_events();
 
-  // See if there are any events remaining to upload
-  if (uploadEvents->size() < events.count) {
-    NSRange nextEventsRange =
-        NSMakeRange(uploadEvents->size(), events.count - uploadEvents->size());
-    NSArray *nextEvents = [events subarrayWithRange:nextEventsRange];
-    return [self uploadEvents:nextEvents];
+  NSUInteger i = 0;
+  while (i < events.count) {
+    SNTStoredEvent *event = events[i++];
```

Comment:
Consider using `enumerateObjectsUsingBlock:`. It'll almost certainly be faster than getting the event out separately and you get the index as you go along.